tachyon/math/polynomials/univariate/radix2_evaluation_domain.h

```diff
@@ -267,6 +363,130 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree> {
+    // |OPENMP_NESTED_PARALLEL_FOR| in |RunDitLayers|
+    for (size_t block_start = 0; block_start < this->size_;
+         block_start += chunk_rows) {
+      size_t cur_chunk_rows = std::min(chunk_rows, this->size_ - block_start);
+      Eigen::Block<RowMajorMatrix<F>> submat =
+          mat.block(block_start, 0, cur_chunk_rows, cols);
+      for (size_t layer = 0; layer < mid_; ++layer) {
+        RunDitLayers(submat, layer, absl::MakeSpan(twiddles), false);
+      }
+    }
+  }
+
+  // This can be used as the second half of a parallelized butterfly network.
+  CONSTEXPR_IF_NOT_OPENMP void RunParallelRowChunksReversed(
+      RowMajorMatrix<F>& mat, const std::vector<F>& twiddles_rev) {
+    size_t cols = static_cast<size_t>(mat.cols());
+    CHECK(base::bits::IsPowerOfTwo(mat.rows()));
+    size_t chunk_rows = 1 << (this->log_size_of_group_ - mid_);
+
+    // max block size: 2^(|this->log_size_of_group_| - |mid_|)
+    // TODO(ashjeong): benchmark between |OPENMP_PARALLEL_FOR| here vs
+    // |OPENMP_NESTED_PARALLEL_FOR| in |RunDitLayers|
+    for (size_t block_start = 0; block_start < this->size_;
+         block_start += chunk_rows) {
+      size_t thread = block_start / chunk_rows;
+      size_t cur_chunk_rows = std::min(chunk_rows, this->size_ - block_start);
+      Eigen::Block<RowMajorMatrix<F>> submat =
+          mat.block(block_start, 0, cur_chunk_rows, cols);
+      for (size_t layer = mid_; layer < this->log_size_of_group_; ++layer) {
+        size_t first_block = thread << (layer - mid_);
+        RunDitLayers(submat, layer,
+                     absl::MakeSpan(twiddles_rev.data() + first_block,
+                                    twiddles_rev.size() - first_block),
+                     true);
+      }
+    }
+  }
+
+  CONSTEXPR_IF_NOT_OPENMP void RunDitLayers(
+      Eigen::Block<RowMajorMatrix<F>>& submat, size_t layer,
+      const absl::Span<const F>& twiddles, bool rev) {
+    size_t layer_rev = this->log_size_of_group_ - 1 - layer;
+    size_t half_block_size = rev ? 1 << layer_rev : 1 << layer;
+    size_t block_size = half_block_size * 2;
+    size_t sub_rows = static_cast<size_t>(submat.rows());
+    DCHECK(sub_rows >= block_size);
+
+    OPENMP_PARALLEL_NESTED_FOR(size_t block_start = 0; block_start < sub_rows;
+                               block_start += block_size) {
+      for (size_t i = 0; i < half_block_size; ++i) {
+        size_t hi = block_start + i;
+        size_t lo = hi + half_block_size;
+        const F& twiddle =
+            rev ? twiddles[block_start / block_size] : twiddles[i << layer_rev];
+        ApplyButterflyToRows(submat, hi, lo, twiddle);
+      }
+    }
+  }
+
+  CONSTEXPR_IF_NOT_OPENMP std::vector<F> ReverseSliceIndexBits(
+      const std::vector<F>& vals) {
+    size_t n = vals.size();
+    if (n == 0) {
+      return vals;
+    }
+    CHECK(base::bits::IsPowerOfTwo(n));
+    size_t log_n = base::bits::SafeLog2Ceiling(n);
```

Comment:
```suggestion
    size_t log_n = base::bits::Log2Ceiling(n);
```
Since it returns early when n is 0, I think we can use `Log2Ceiling` here.